mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java

```diff
@@ -0,0 +1,571 @@
+      if (inMemoryDataModel == InMemoryDataModel.DEFAULT) {
+        Expression filter = currentTask.residual();
+        return applyResidualsOnGenericRecords(iterable, filter).iterator();
+      } else {
+        return iterable.iterator();
+      }
+    }
+
+    private Iterable<T> applyResidualsOnGenericRecords(CloseableIterable<T> iterable, Expression filter) {
+      if (filter == null || filter.equals(Expressions.alwaysTrue())) {
+        return iterable;
+      } else {
+        throw new UnsupportedOperationException(
+            String.format("Filter expression %s is not completely satisfied. Additional rows can be returned" +
+                              " not satisfied by the filter expression", filter));
+      }
+    }
+
+    @SuppressWarnings("unchecked")
+    private T withPartitionColumns(T row, Schema identityPartitionSchema, PartitionSpec spec, StructLike partition) {
+      switch (inMemoryDataModel) {
+        case PIG:
+        case HIVE:
+          // TODO implement adding partition columns to records for Pig and Hive
+          throw new UnsupportedOperationException();
+        case DEFAULT:
+          return (T) genericRecordWithPartitionsColumns((Record) row, identityPartitionSchema, spec, partition);
+      }
+      return row;
+    }
+
+    private static Record genericRecordWithPartitionsColumns(
+        Record record, Schema identityPartitionSchema, PartitionSpec spec, StructLike partition) {
+      List<Types.NestedField> fields = Lists.newArrayList(record.struct().fields());
+      fields.addAll(identityPartitionSchema.asStruct().fields());
+      GenericRecord row = GenericRecord.create(Types.StructType.of(fields));
+      int size = record.struct().fields().size();
+      for (int i = 0; i < size; i++) {
+        row.set(i, record.get(i));
+      }
+      List<PartitionField> partitionFields = spec.fields();
+      List<Types.NestedField> identityColumns = identityPartitionSchema.columns();
+      for (int i = 0; i < identityColumns.size(); i++) {
+        Types.NestedField identityColumn = identityColumns.get(i);
+
+        for (int j = 0; j < partitionFields.size(); j++) {
+          PartitionField partitionField = partitionFields.get(j);
+          if (identityColumn.fieldId() == partitionField.sourceId() &&
+              "identity".equals(partitionField.transform().toString())) {
+            //TODO: identity partitions are being added to the end, this changes
+            // the position of the column. This seems like a potential bug
+            row.set(size + i, partition.get(j, spec.javaClasses()[i]));
+          }
+        }
+      }
+      return row;
+    }
+
+    private CloseableIterable<T> newAvroIterable(InputFile inputFile, FileScanTask task, Schema readSchema) {
+      Avro.ReadBuilder avroReadBuilder = Avro.read(inputFile)
+                                             .project(readSchema)
+                                             .split(task.start(), task.length());
```

Comment:
Nit: typical style is to use 2 indents / 4 spaces from the next line, not to indent to the same level as the last function call.